### PR TITLE
Run containers as a non-root user

### PR DESCRIPTION
**Pull request by alexintech**
# What this PR does

Create a custom non-root user and use it to start an app. So uwsgi does not require to use `setUid` and `setGid` system calls.

It handles errors while starting in Kubernetes with `runAsNonRoot: true` check. 

## Which issue(s) this PR fixes

closes https://github.com/grafana/oncall/issues/445

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [ ] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)

### CODE REVIEW

File: engine/Dockerfile
Comment (joeyorlando) [inline engine/Dockerfile:5]: These flags are equivalent according to the [manpage](https://man7.org/linux/man-pages/man8/groupadd.8.html#:~:text=The%20groupadd%20command%20creates%20a,digits%2C%20underscores%2C%20or%20dashes.) for `groupadd`
```suggestion
RUN groupadd --system --gid 1000 ${APP_USER} && \
```
I would prefer the more verbose flags for readability

File: engine/Dockerfile
Comment (joeyorlando) [inline engine/Dockerfile:6]: same suggestion here. also, why the `--no-log-init` flag?
```suggestion
    useradd --no-log-init --system --uid 1000 --gid ${APP_USER} ${APP_USER}
```

File: engine/Dockerfile
Comment (joeyorlando) [inline engine/Dockerfile:1]: consider deduplicating `1000` to an `ENV` variable at the top of the `Dockerfile` so that it can be reused on lines 5, 6, and 57

File: docker-compose-developer.yml
Comment (joeyorlando) [inline docker-compose-developer.yml:7]: is this needed? according to the `docker-compose` configuration [docs](https://docs.docker.com/compose/compose-file/05-services/#user):

> `user` overrides the user used to run the container process. Default is that set by image (i.e. Dockerfile `USER`), if not set, `root`. 

You are setting `USER` in `engine/Dockerfile` [here](https://github.com/grafana/oncall/pull/2053/files#diff-a25c2322baf692c37d4ecc9c3611772e78dfb8bbaf9cc4ef27b507e60cb3ddc9R57), so I think this may be redundant

**Review by joeyorlando — COMMENTED** [review]
Overall looks great to me! Thanks for picking this up. Will approve once you've addressed the questions/concerned listed 😄 

@iskhakov do you mind giving this a second set of 👀 ?

File: docker-compose-developer.yml
Comment (alexintech) [reply docker-compose-developer.yml:58]: When I run `make init`, `node_modules` folder on my host machine is empty and PyCharm cannot correctly understand the project. Or maybe I don't know how to tell PyCharm where to find node_modules?

I can revert these changes if they break developer experience 😄

File: engine/Dockerfile
Comment (alexintech) [reply engine/Dockerfile:1]: I considered deduplicate `1000`, but how `ENV` variable should be named? `APP_USER_UID_GID`? Logically there should be two variables `APP_USER_UID` and `APP_USER_GID`.

I didn't want to introduce extra `ENV` variables for the value, that used only 3 times and should not be reused again in the `Dockerfile`

File: engine/Dockerfile
Comment (alexintech) [reply engine/Dockerfile]: I don't remember where I found the recommendation to use `--no-log-init`. With this flag `useradd` does not write to `/var/log/faillog` and `/var/log/lastlog` and saves around `300Kb` 😄 It's not a lot but I think it's not a bad flag for containerized user.

**Review by iskhakov — CHANGES_REQUESTED** [review]
overall looks good to me

I would like to make a final look before merging, so I leave my review in "Request changes" state

File: docker-compose-developer.yml
Comment (joeyorlando) [reply docker-compose-developer.yml:7]: ahh yes, that makes sense. thanks for the detailed explanation here 👍

File: docker-compose-developer.yml
Comment (joeyorlando) [reply docker-compose-developer.yml:58]: > `node_modules` folder on my host machine is empty

this is correct. [`make init`](https://github.com/grafana/oncall/blob/dev/Makefile#L91) spins up an ephemeral `oncall_ui` container, defined in `docker-compose-dev.yml`, and runs `yarn install && yarn build:dev`.

This installs `node_modules` to the `node_modules_dev` volume mount, not `./grafana-plugin/node_modules` on your host machine & builds the plugin dist to `./grafana-plugin/dist` (on your host-machine). There's more context in the PR description of #1890 on this change. tldr; now you no longer need to have `node` installed on your machine to worry about installing these deps.

Comment (joeyorlando) [reply docker-compose-developer.yml:58]: re: PyCharm, you'd have to point it to the volume mounted `node_modules`, which I suspect is not possible? what if you ran `yarn install` on your host machine (on top of `make init`), does this fix it?

Comment (alexintech) [reply docker-compose-developer.yml:58]: > what if you ran yarn install on your host machine (on top of make init), does this fix it?

Yes, that fixes. But in this case there are `node_modules` in `grafana_plugin/node_modules` and in a docker volume. And you're also need installed `node` to run `yarn install` 😄 

> Ideally we store the node_modules only once inside the named volumes.

I only do not understand, why `node_modules` inside the named volume is better than in `grafana_plugin/node_module` on host machine?

Comment (joeyorlando) [reply docker-compose-developer.yml:58]: > why node_modules inside the named volume is better than in grafana_plugin/node_module on host machine

in some sense it's opinionated, but by storing them on the volume, you don't require the developer to have `node` installed on their host machine (because the container takes care of this for you). The frontend will only run w/ specific versions of `node`, so this just makes it one less thing the dev needs to worry about 😄

Comment (alexintech) [reply docker-compose-developer.yml:58]: > you don't require the developer to have node installed on their host machine (because the container takes care of this for you).

Exactly the same in this case. You run `make init` and it downloads `node_modules` and exposes them to host machine. So PyCharm can find imports and you can dive into sources of installed dependencies if you need. And I suppose other IDEs also may benefit.

Actually, there is no need to mount  `./grafana-plugin/node_modules:/etc/app/node_modules`, as `./grafana-plugin:/etc/app` already can expose node_modules.

What code editor do you use?

Comment (joeyorlando) [reply docker-compose-developer.yml:58]: lets move this discussion to a separate GitHub issue so that we don't block the other work you've done here 😄 I've created #2097.

Comment (alexintech) [reply docker-compose-developer.yml]: Sure!

File: engine/Dockerfile
Comment (iskhakov) [inline engine/Dockerfile:5]: can we keep gid=2000 to make sure it is backwards-compatible?

Comment (alexintech) [reply engine/Dockerfile:5]: Yes, we can change to `gid=2000`. 

I think it's a common practice to set the same `uid` and `gid` for a custom user. I suppose that nothing would break, as process that starts application have the same user permissions. Something can be broken, if another process with some other uid and in the same group wants to use files with group permissions. But I suppose it's not applicable to dockerized application.

I'll change to `gid=2000` if you prefer that.

Comment (alexintech) [reply engine/Dockerfile:5]: Hi @iskhakov @joeyorlando ! Change to `gid=2000` or not ?

Comment (joeyorlando) [reply engine/Dockerfile:5]: I agree w/ @iskhakov , lets keep it to 2000 just to be safe

**Review by iskhakov — APPROVED** [review]

File: engine/Dockerfile
Comment (joeyorlando) [inline engine/Dockerfile:59]: @alexintech actually one thought came to mind. If we move `USER 1000` higher up to line 13, I _think_ we may not need any of the `chown` commands?

Comment (alexintech) [reply engine/Dockerfile:59]: > If we move USER 1000 higher up to line 13

That will break `RUN apt-get install -y sqlite3 default-mysql-client postgresql-client` [here](https://github.com/grafana/oncall/pull/2053#diff-a25c2322baf692c37d4ecc9c3611772e78dfb8bbaf9cc4ef27b507e60cb3ddc9R49) in `dev` target.

Comment (joeyorlando) [reply engine/Dockerfile:59]: that makes sense 👍 

I opened [a PR](https://github.com/grafana/oncall/pull/2132) to move `USER` to the end of the `dev` target so that it is present if a different build `target` is specified to `docker build`

**Pull request merged**
